py-pyseer and py-glmnet-python: new packages

luke-dt opened the pull request.

luke-dt committed:
initial commit

luke-dt committed:
first try

just added deps as they are listed. have not built yet

luke-dt committed:
builds, but fails some tests

tests/enet_test.py::TestFitEnet::test_fit_binary FAILED                                             [ 13%]
tests/enet_test.py::TestFitEnet::test_fit_continuous FAILED                                         [ 15%]

luke-dt committed:
fixed glmnet conflict with newer scipy/numpy

luke-dt committed:
builds, added bwa

dependencies are a mess, but still installs incorrect py-scipy version

luke-dt committed:
fixed bedtools and dendropy version numbers

luke-dt committed:
fixed config.yaml for upstream

luke-dt committed:
fix style

tldahlgren reviewed (dismissed):
Confirmed the homepages and version sha256, 'spack spec' succeeds.

tldahlgren commented:
Adding @adamjstewart and @pradyunsg as reviewers and will defer merge to one or both of them.

manuelakuhn commented on var/spack/repos/builtin/packages/py-pyseer/package.py:
```diff
@@ -0,0 +1,47 @@
+    depends_on("py-scikit-learn@0.20.0:", type=("build", "run"))
+    depends_on("py-statsmodels@0.9.0:", type=("build", "run"))
+    depends_on("py-pysam@0.15.3:", type=("build", "run"))
+    depends_on("py-glmnet-python@1.0", type=("build", "run"))
```
According to [setup.py#L78](https://github.com/mgalardini/pyseer/blob/1.3.11/setup.py#L78) this is the wrong dependency. It is not the repo from bbalasub1 but the one from johnlees (which is a fork from bbalasub1) which is used. In addition it is version 1.0.2.

@adamjstewart there seem to be quite some glmnet-python repos and packages out there. I think just calling the first one we add to spack `glmnet-python` can be confusing in case another package will need one of the others in the future. Any suggestions on the name or do you think it is not a problem?

luke-dt replied:
haha, incredible. It looks like pip and conda both download from the johnlees repo, but then all the homepages and source code link to the bbalasub1 repo, which is how I got mixed up:

https://anaconda.org/conda-forge/glmnet_py
https://github.com/conda-forge/glmnet_py-feedstock/blob/a32dd372c6e1a8cc25cd9dfd224777cebb15ff4b/recipe/meta.yaml#L2

I was wondering where conda-forge got v1.0.2 from. I'll make this change and have to confirm that it still works as expected.

manuelakuhn replied:
Yes, I also found it quite confusing. What makes it even worse is that both the bbalasub1 and johnlees repos claim that the packages can be installed with `pip install glmnet_py` but when you check that on pypi this is again an additional repo (namely hanfang) ...

manuelakuhn commented on var/spack/repos/builtin/packages/py-pyseer/package.py:
```diff
@@ -0,0 +1,47 @@
+    depends_on("py-scipy@1.1.0:", type=("build", "run"))
+    depends_on("py-pandas@0.23.4:", type=("build", "run"))
+    depends_on("py-scikit-learn@0.20.0:", type=("build", "run"))
+    depends_on("py-statsmodels@0.9.0:", type=("build", "run"))
```
```suggestion
    depends_on("py-statsmodels@0.10.0:", type=("build", "run"))
```
see [setup.py#71](https://github.com/mgalardini/pyseer/blob/1.3.11/setup.py#L71)

manuelakuhn commented on var/spack/repos/builtin/packages/py-glmnet-python/package.py:
```diff
@@ -0,0 +1,23 @@
+    depends_on("py-setuptools", type="build")
+    depends_on("py-joblib@0.10.3:", type=("build", "run"))
+    # https://github.com/bbalasub1/glmnet_python/issues/64
+    depends_on("py-scipy@:1.8.1", type=("build", "run"))
```
I think you built a package for the wrong dependency here. See below.

manuelakuhn commented on var/spack/repos/builtin/packages/py-pyseer/package.py:
```diff
@@ -0,0 +1,47 @@
+    # optional deps
+    depends_on("py-matplotlib@2.1.0:", when="+matplotlib", type=("build", "run"))
+    depends_on("py-dendropy@4.3.0:", when="+dendropy", type=("build", "run"))
+    depends_on("py-pybedtools@0.7.10:", when="+kmer", type=("build", "run"))
```
`py-matplotlib`, `py-dendropy` and `py-pybedtools` are mandatory dependencies according to `setup.py`
I would remove the variants for them then.

In addition `py-dendropy@4.4.0:` is already defined as mandatory above. So adding it here in a lesser version is redundant.

luke-dt replied:
oh, the duplicate dendropy dependency was a typo. It's used for the scripts/phylogeny_distance.py util, but not the main pyseer function so I included it as a recommended variant. I can remove that and make it required it that's simpler. Same with matplotlib (used in a couple utils)

manuelakuhn replied:
As far as I know all dependencies listed in `install_requires` in `setup.py` are required dependencies in spack (If I understood it correctly otherwise it tries to download them). That means installing with one of these variants disabled would result in missing dependencies. So I would just add them as normal dependencies and remove the variants.

manuelakuhn commented on var/spack/repos/builtin/packages/py-pyseer/package.py:
```diff
@@ -0,0 +1,47 @@
+    depends_on("py-dendropy@4.3.0:", when="+dendropy", type=("build", "run"))
+    depends_on("py-pybedtools@0.7.10:", when="+kmer", type=("build", "run"))
+    depends_on("bedops@2.4.9:", when="+kmer", type=("build", "run"))
+    depends_on("bwa", when="+kmer", type=("build", "run"))
```
I am a little confused on which basis which dependency was added: 

According to [installation.html](https://pyseer.readthedocs.io/en/master/installation.html) `bedtools` is also a dependency. I guess it was not explicitly defined because it is an implicit dependency of `py-pybedtools`. But is `bedtools2` (the one from `py-pybedtools`) the same as `bedtools` from here`?

On which basis did you choose additional dependencies from [bioconda](https://github.com/bioconda/bioconda-recipes/blob/master/recipes/pyseer/meta.yaml). E.g. you have added `bwa` but not `mash` or `python-dateutil`.

luke-dt replied:
I didn't include `bedtools` because it's an implicit dependency of `py-pybedtools` and it seemed redundant to list it twice. I can add it back it that's better.

`bwa` is used in a a util (`pyseer/kmer_mapping/bwa.py`). 

`mash` is an unnecessary dependency. It would be useful to have in the same environment as pyseer can interact with its output, but the executable is never directly called by `pyseer`.[ In the `pyseer` tutorial, they just run mash directly from the command line and then use a `pyseer` function on the output.](https://github.com/mgalardini/pyseer/blob/master/docs/tutorial.rst). Again, I can add it if you feel that is better.

I have no idea why `python-dateutil` is listed as a dependency. It's not used in the `pyseer` source code at all beyond the requirements list. I have confirmed that it runs fine without it.

luke-dt replied:
Oh, and bedtools and bedtools2 are the same thing. Since v2, bedtools uses a new repo called "bedtools2" but the name in bioconda is still "bedtools".

https://github.com/arq5x/bedtools2

https://github.com/arq5x/bedtools

https://github.com/bioconda/bioconda-recipes/blob/master/recipes/bedtools/meta.yaml

manuelakuhn replied:
Regarding bedtools. I don't know how the policy is for implicit dependencies in spack. I was just asking because it is explicitly listed in the installation description. Maybe @adamjstewart can help here.

For the other dependencies, I think the lesser the better as long as the software still works. Maybe adding comments would help.

Ah, I also just saw that this is also described in the [README.md](https://github.com/mgalardini/pyseer/blob/master/README.md#prerequisites).
So in total all you took from bioconda was the `pysam` version, right? So maybe I would refer to `installation.html` and `README.md` for the dependencies in your comment above and just make a small comment to `pysam` about the version. But this is a little bit of nitpicking here :-)

luke-dt committed:
update glmnet-python to new repo

manuelakuhn commented on var/spack/repos/builtin/packages/py-glmnet-python/package.py:
```diff
@@ -0,0 +1,25 @@
+
+    depends_on("py-setuptools", type="build")
+    depends_on("py-joblib@0.10.3:", type=("build", "run"))
+    depends_on("py-scipy", type=("build", "run"))
```
Maybe you can add a comment here where this dependency is coming from because it is not listed in the `setup.py`

Other than this, `py-glmnet-python` looks good to me now.

luke-dt committed:
removed variants, added better comments

adamjstewart commented on var/spack/repos/builtin/packages/py-pyseer/package.py:
```diff
@@ -0,0 +1,47 @@
+    depends_on("py-dendropy@4.4.0:", type=("build", "run"))
+    depends_on("py-tqdm@4.20.0:", type=("build", "run"))
+
+    # optional deps
```
Personally I would make these always required. That's how they're listed in `setup.py`

adamjstewart commented on var/spack/repos/builtin/packages/py-pyseer/package.py:
```diff
@@ -0,0 +1,47 @@
+    # https://pyseer.readthedocs.io/en/master/installation.html
+    # also from
+    # github.com/bioconda/bioconda-recipes/blob/master/recipes/pyseer/meta.yaml
+    # the deps are messy and this project is not updated
```
A lot of these version requirements differ from setup.py, docs, and conda. Where are they coming from?

adamjstewart reviewed (commented):
Completely agree with the naming decisions here, forks are fine as long as they don't have conflicting names.

luke-dt committed:
improved documentation, confirmed runs

-added comments to clarify where different versions came from
-confirmed builds and runs with real data

luke-dt committed:
Merge branch 'develop' into package/py-pyseer

adamjstewart commented on var/spack/repos/builtin/packages/py-pyseer/package.py:
```diff
@@ -0,0 +1,43 @@
+    depends_on("py-numpy@1.15.2:", type=("build", "run"))
+    depends_on("py-scipy@1.1.0:", type=("build", "run"))
+    depends_on("py-pandas@0.23.4:", type=("build", "run"))
+    depends_on("py-scikit-learn@0.20.0:", type=("build", "run"))
```
These lower bounds are just the versions that were tested, they aren't necessarily the versions that are required. I would remove these constraints.

adamjstewart reviewed (commented):
I would keep lower bounds only when setup.py or meta.yaml use them, not for all dependencies.

luke-dt commented:
packages added in #44543
